### DISCUSSION TITLE
Speed up git-gui when using sparse-index

### DISCUSSION THREAD

**Discussion by Isengart** [microsoft/git#530] [General]
We have a big monorepo and therefore using sparse checkout and sparse index.
I noticed that using git-gui is still a pain.
Each click between changed files takes ~4-5s on my machine until the diff is shown.

I analyzed this and tracked it down to that the following commands are not yet activated to use the sparse index (command_requires_full_index is active for them):
* git diff-files
* git check-attr
* git diff-index (for staged changes)

I build git myself and disabled the command_requires_full_index setting.
Of course git-gui is then much more responsive.

I wonder what the needed steps would be to get activation for these commands merged.
I assume there need to be testcases?

For git diff-files I think one could argue that the code might get quickly into the same pathes as git diff for which the sparse index is already activated.

**Comment by derrickstolee** [comment]
Hi @Isengart. Thanks for doing the hard work of tracking down which plumbing commands are used by Git GUI. You are right that custom-compiling to stop the sparse index expansion in those commands will fix the performance issue, there might be some corner cases that are broken.

The biggest thing that is necessary for knowing if everything works is to add tests for those cases. The test file [`t1092-sparse-checkout-compatibility.sh`](https://github.com/git/git/blob/HEAD/t/t1092-sparse-checkout-compatibility.sh) contains the tests that ensure things work the same with and without the sparse index, as well as checking that things will use the fast path when possible.

@ffyuanda is a Google Summer of Code student currently working on sparse index integrations. I have shared this discussion with him to see if he has time to approach these commands during his supported time.
